packages/commonwealth/server/routes/threads/create_thread_comment_handler.ts

```diff
@@ -58,43 +69,49 @@ export const createThreadCommentHandler = async (
     // @ts-expect-error <StrictNullChecks>
     address,
     parentId,
-    // @ts-expect-error <StrictNullChecks>
-    threadId: parseInt(threadId, 10) || undefined,
+    threadId: parseInt(threadId, 10),
     text,
     discordMeta: discord_meta,
   };
 
   if (hasCanvasSignedDataApiArgs(req.body)) {
     threadCommentFields.canvasSignedData = req.body.canvas_signed_data;
-    threadCommentFields.canvasHash = req.body.canvas_hash;
-
-    if (config.ENFORCE_SESSION_KEYS) {
-      const { canvasSignedData } = fromCanvasSignedDataApiArgs(req.body);
-      await verifyComment(canvasSignedData, {
-        thread_id: parseInt(threadId, 10) || undefined,
-        text,
-        address:
-          canvasSignedData.actionMessage.payload.address.split(':')[0] ==
-          'polkadot'
-            ? addressSwapper({
-                currentPrefix: 42,
-                // @ts-expect-error <StrictNullChecks>
-                address: address.address,
-              })
-            : // @ts-expect-error <StrictNullChecks>
-              address.address,
-        parent_comment_id: parentId,
-      });
-    }
+    threadCommentFields.canvasMsgId = req.body.canvas_msg_id;
+
+    const { canvasSignedData } = fromCanvasSignedDataApiArgs(req.body);
+    const canvasComment = {
+      thread_id: threadMsgId ?? null,
+      text,
+      address:
+        canvasSignedData.actionMessage.payload.did.split(':')[2] == 'polkadot'
+          ? addressSwapper({
+              currentPrefix: 42,
+              // @ts-expect-error <StrictNullChecks>
+              address: address.address,
+            })
+          : // @ts-expect-error <StrictNullChecks>
+            address.address,
+      parent_comment_id: parentCommentMsgId ?? null,
+    };
+    await verifyComment(canvasSignedData, canvasComment);
   }
 
+  // create thread comment
   const [comment, notificationOptions, analyticsOptions] =
     await controllers.threads.createThreadComment(threadCommentFields);
 
+  // publish signed data
+  if (hasCanvasSignedDataApiArgs(req.body)) {
+    const { canvasSignedData } = fromCanvasSignedDataApiArgs(req.body);
+    await applyCanvasSignedData(canvas, canvasSignedData);
```

Comment:
Since these are being written to a separate database, they are not able to be in the same transaction so we have a possibility of data inconsistency where the federation db is out of sync with the commonwealth db. I don't really understand the purpose of the federation db, but do we have some plan on how we will reconcile them in the future?

Reply:
We need some kind of persistent data store (either Postgres or block storage/disk) to maintain a copy of the Merkle tree used for Canvas federation. Currently a separate Postgres db is the easiest way to get that.

It should probably be a separate database from Commonwealth, so that in case an external user pushes a bunch of data to it, that doesn't end up impacting Commonwealth uptime or or storage caps, and we have a clear boundary where we can implement rate limits.